services/vision/obstacledepth/obstacle_depth.go

```diff
@@ -0,0 +1,377 @@
+	defer span.End()
+	if conf == nil {
+		return nil, errors.New("config for obstacle_depth cannot be nil")
+	}
+
+	// If you have no intrinsics
+	if conf.Intrinsics == nil {
+		r.Logger().Warn("obstacle depth started without camera's intrinsic parameters")
+		segmenter := buildObsDepthNoIntrinsics()
+		return svision.NewService(name, r, nil, nil, nil, segmenter)
+	}
+
+	// Use defaults if needed
+	if conf.Hmax == 0 {
+		conf.Hmax = defaultHmax
+	}
+	if conf.ThetaMax == 0 {
+		conf.ThetaMax = defaultThetamax
+	}
+	if conf.K == 0 {
+		conf.K = defaultK
+	}
+
+	myObsDep := obsDepth{
+		hMin: conf.Hmin, hMax: conf.Hmax, sinTheta: math.Sin(conf.ThetaMax),
+		intrinsics: conf.Intrinsics, returnPCDs: conf.ReturnPCDs, k: conf.K,
+	}
+	segmenter := myObsDep.buildObsDepthWithIntrinsics() // does the thing
+	return svision.NewService(name, r, nil, nil, nil, segmenter)
+}
+
+// buildObsDepthNoIntrinsics will return the shortest depth in the depth map as a Geometry point.
+func buildObsDepthNoIntrinsics() segmentation.Segmenter {
+	return func(ctx context.Context, src camera.VideoSource) ([]*vision.Object, error) {
+		depthStream, err := src.Stream(ctx)
+		if err != nil {
+			return nil, errors.Errorf("could not get stream from %s", src)
+		}
+		pic, release, err := depthStream.Next(ctx)
+		if err != nil {
+			return nil, errors.Errorf("could not get image from stream %s", depthStream)
+		}
+		defer release()
+		dm, err := rimage.ConvertImageToDepthMap(ctx, pic)
+		if err != nil {
+			return nil, errors.New("could not convert image to depth map")
+		}
+		min, _ := dm.MinMax()
+
+		pt := spatialmath.NewPoint(r3.Vector{X: 0, Y: 0, Z: float64(min)}, "")
+		toReturn := make([]*vision.Object, 1)
+		toReturn[0] = &vision.Object{Geometry: pt}
+
+		return toReturn, nil
+	}
+}
+
+// buildObsDepthWithIntrinsics will use the methodology in Manduchi et al. to find obstacle points
+// before clustering and projecting those points into 3D obstacles.
+func (o *obsDepth) buildObsDepthWithIntrinsics() segmentation.Segmenter {
+	return func(ctx context.Context, src camera.VideoSource) ([]*vision.Object, error) {
+		// Check if we have intrinsics here or in the camera properties. If not, don't even try
+		props, err := src.Properties(ctx)
+		if err != nil {
+			return nil, err
+		}
+		if props.IntrinsicParams != nil {
+			o.intrinsics = props.IntrinsicParams
+		}
+		if o.intrinsics == nil {
+			return nil, errors.New("tried to build obstacles depth with intrinsics but no instrinsics found")
+		}
+
+		depthStream, err := src.Stream(ctx)
+		if err != nil {
+			return nil, errors.Errorf("could not get stream from %s", src)
+		}
+		pic, release, err := depthStream.Next(ctx)
+		if err != nil {
+			return nil, errors.Errorf("could not get image from stream %s", depthStream)
+		}
+		defer release()
+		dm, err := rimage.ConvertImageToDepthMap(ctx, pic)
+		if err != nil {
+			return nil, errors.New("could not convert image to depth map")
+		}
+		o.dm = dm
```

Comment:
I think there are more loop through the depthMap than are necessary. it's best to reduce the amount of copies of the points. Try this nested for loop that doesn't use go-rountines. Is there any difference in speed than what you currently have?
```
w, h := dm.Width(), dm.Height()
obstaclePoints := []image.Point{}
for i := 0; i < w; i += n { // do the sub-sampling
  for j := 0; j < h; j++ {
    candidate := image.Pt(i, j)
    // for every sub-sampled point, figure out if it is compatible with being an obstacle
    for l := 0; l < w; l += n  {  // continue with the sub-sampling
      for m := 0; m < h; m++ {
        compareTo := image.Pt(l, m)
     	if candidate == compareto {
          continue
        }
        if o.isCompatible(candidate, compareTo) {
          obstaclePoints = append(obstaclePoints, candidate)
        }
     }
  } 
}   
```

Reply:
If parallelization needs to be done, we can create a channel that you return obstacle Points on, like in the comment below. Rather than creating and looping over a list of points, you can just make it so each go routine loops over a different subset of candidates. essentially...
```
for i := 0; i < w; i += n { 
  for j := 0; j < h; j++ {
    // only loop over fraction of points that the go rountine needs to
    // nGoRoutines are the total number of goRountines that are running. each rountine gets an id myID,
    // which is a number between [0, nGoRoutines-1].
    if (j % nGoRoutines) != myID {
       continue
    }
    candidate := image.Pt(i, j)
    // for every sub-sampled point, figure out if it is compatible with being an obstacle
    for l := 0; l < w; l += n  {  // continue with the sub-sampling
      for m := 0; m < h; m++ {
        compareTo := image.Pt(l, m)
     	if candidate == compareto {
          continue
        }
        if o.isCompatible(candidate, compareTo) {
          obstaclePointsChan <-candidate
        }
     }
  } 
} 
```